File: apina-core/src/test/java/fi/evident/apina/java/reader/ClassMetadataReaderTest.java
package fi.evident.apina.java.reader;

import fi.evident.apina.java.model.JavaClass;
import fi.evident.apina.java.model.JavaField;
import org.junit.Test;

import java.util.Comparator;
import java.util.List;

import static fi.evident.apina.java.reader.ClassReaderUtils.loadClass;
import static fi.evident.apina.java.reader.JavaTypeMatchers.*;
import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

public class ClassMetadataReaderTest {

    @Test
    public void loadingFields() {
        JavaClass javaClass = loadClass(TestClass.class);

        List<JavaField> fields = javaClass.getFields();

        assertThat(javaClass.getSchema(), is(singletonSchema("T", basicType(CharSequence.class))));

        assertThat(fields.size(), is(3));
        assertThat(javaClass.getField("field1").getType(), is(typeWithRepresentation("java.lang.String")));
        assertThat(javaClass.getField("field2").getType(), is(typeWithRepresentation("java.util.List<java.lang.String>")));
        assertThat(javaClass.getField("field3").getType(), is(typeVariable("T")));
    }

    @Test
    public void innerClassWithOuterBounds() {
        loadClass(AnonymousInnerClassWithOuterBounds.createInnerClassInstance().getClass());
    }

    @Test
    public void enumClasses() {
        JavaClass javaClass = loadClass(TestEnum.class);

        assertThat(javaClass.isEnum(), is(true));
        assertThat(javaClass.getEnumConstants(), is(asList("FOO", "BAR", "BAZ")));
    }

    @SuppressWarnings("unused")
    private enum TestEnum {
        FOO, BAR, BAZ;

        public String instanceField;
        public static String staticField;
    }

    @SuppressWarnings("unused")
    private static final class TestClass<T extends CharSequence> {

        public String field1;

        public List<String> field2;

        private T field3;

        public void method1() {
        }

        public String method2() {
            throw new UnsupportedOperationException();
        }

        public T method3(T x) {
            throw new UnsupportedOperationException();
        }
    }

    private static final class AnonymousInnerClassWithOuterBounds {

        @SuppressWarnings("Convert2Lambda")
        public static <T> Comparator<T> createInnerClassInstance() {
            return new Comparator<T>() {
                @SuppressWarnings("ComparatorMethodParameterNotUsed")
                @Override
                public int compare(T o1, T o2) {
                    return 0;
                }
            };
        }
    }
}

File: apina-core/src/test/kotlin/fi/evident/apina/java/reader/ClassMetadataReaderTest.kt
package fi.evident.apina.java.reader

import fi.evident.apina.java.reader.JavaTypeMatchers.basicType
import fi.evident.apina.java.reader.JavaTypeMatchers.singletonSchema
import fi.evident.apina.java.reader.JavaTypeMatchers.typeVariable
import fi.evident.apina.java.reader.JavaTypeMatchers.typeWithRepresentation
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.Test
import java.util.*
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

@Suppress("UNUSED_PARAMETER")
class ClassMetadataReaderTest {

    @Test
    fun `loading fields`() {
        val javaClass = loadClass(TestClass::class.java)

        val fields = javaClass.fields

        assertThat(javaClass.schema, singletonSchema("T", basicType(CharSequence::class.java)))

        assertEquals(3, fields.size)
        assertThat(javaClass.getField("field1").type, typeWithRepresentation("java.lang.String"))
        assertThat(javaClass.getField("field2").type, typeWithRepresentation("java.util.List<java.lang.String>"))
        assertThat(javaClass.getField("field3").type, typeVariable("T"))
    }

    @Test
    fun `inner class with outer bounds`() {
        loadClass(AnonymousInnerClassWithOuterBounds.createInnerClassInstance<Any>().javaClass)
    }

    @Test
    fun `enum classes`() {
        val javaClass = loadClass(TestEnum::class.java)

        assertTrue(javaClass.isEnum)
        assertEquals(listOf("FOO", "BAR", "BAZ"), javaClass.enumConstants)
    }

    @Test
    fun `kotlin metadata`() {
        val metadata = loadClass(TestKotlinClass::class.java).kotlinMetadata
        assertNotNull(metadata)

        assertEquals(setOf("foo", "bar"), metadata.properties.map { it.name }.toSet())
    }

    @Suppress("unused")
    private class TestKotlinClass {
        val foo = ""
        val bar = ""
    }

    @Suppress("unused")
    private enum class TestEnum {
        FOO, BAR, BAZ;

        var instanceField: String? = null

        companion object {
            var staticField: String? = null
        }
    }

    @Suppress("unused", "UNUSED_PARAMETER")
    private class TestClass<T : CharSequence> {

        var field1: String? = null

        var field2: List<String>? = null

        private val field3: T? = null

        fun method1() {}

        fun method2(): String = throw UnsupportedOperationException()

        fun method3(x: T): T = throw UnsupportedOperationException()
    }

    private object AnonymousInnerClassWithOuterBounds {
        @Suppress("ObjectLiteralToLambda")
        fun <T> createInnerClassInstance(): Comparator<T> = object: Comparator<T> {
            override fun compare(o1: T, o2: T) = 0
        }
    }
}
